shared/network/graphql/src/commonMain/kotlin/app/futured/kmptemplate/network/graphql/client/ApolloApiAdapter.kt

```diff
@@ -94,16 +98,47 @@ internal class ApolloApiAdapter(
     }
 
     /**
-     * Unwraps successful [ApolloResponse] either into [DATA], or throws [NetworkError.CloudError]
-     * in case errors are present in the response.
+     * https://www.apollographql.com/docs/kotlin/essentials/errors#truth-table
+     * Unwraps successful [ApolloResponse] either into [DATA], or throws [NetworkError].
+     * If the response contains an exception, it throws the exception.
+     * If the response contains complete data, it returns the data.
+     * If the response contains partial data, it returns the data or throws an exception (It's possible to have both
+     * [ApolloResponse.data] non-null and [ApolloResponse.errors] non-null.)
+     * If the response contains errors, it throws [NetworkError.CloudError].
+     * @param throwOnPartialData whether to throw an exception on partial data
      */
-    private fun <DATA : Operation.Data> unfoldResult(response: ApolloResponse<DATA>): DATA =
-        if (response.hasErrors().not()) {
-            response.data ?: error("Response without errors and with no data")
-        } else {
+    private fun <DATA : Operation.Data> unfoldResult(
+        response: ApolloResponse<DATA>,
+        throwOnPartialData: Boolean,
+    ): DATA {
+        response.exception?.let { throw it }
+
+        response.data?.let { data ->
+            if (response.hasErrors()) {
+                if (throwOnPartialData) {
+                    throw NetworkError.CloudError(
+                        code = errorResponseParser.getErrorResponseCode(response),
+                        message = errorResponseParser.getErrorMessage(response),
+                    )
+                }
+                // Return partial data
+                return data
+            }
+            // Return complete data
+            return data
+        }
+
+        if (response.hasErrors()) {
             throw NetworkError.CloudError(
                 code = errorResponseParser.getErrorResponseCode(response),
-                message = errorResponseParser.getErrorMessage(response) ?: "",
+                message = errorResponseParser.getErrorMessage(response),
             )
         }
+
+        // Non-compliant server, if data is null, there should be at least one error
+        throw NetworkError.CloudError(
+            code = errorResponseParser.getErrorResponseCode(response),
+            message = errorResponseParser.getErrorMessage(response),
+        )
+    }
```

Comment:
This two throws are duplicated and can be merged into single throw
```kotlin
// Non-compliant server, if data is null, there should be at least one error, or
// A GraphQL request error happened or a Graph field error bubbled up.
throw NetworkError.CloudError(
    code = errorResponseParser.getErrorResponseCode(response),
    message = errorResponseParser.getErrorMessage(response)
)
```

Reply:
Agreed ✅